File: src/main/java/dev/slne/discord/message/EmbedManager.java
package dev.slne.discord.message;

import dev.slne.discord.exception.ticket.UnableToGetTicketNameException;
import dev.slne.discord.ticket.Ticket;
import dev.slne.discord.ticket.TicketChannelHelper;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import lombok.experimental.ExtensionMethod;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.IMentionable;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

@Service
@ExtensionMethod({TimeFormatter.class})
public class EmbedManager {

  private final TicketChannelHelper ticketChannelHelper;

  public EmbedManager(TicketChannelHelper ticketChannelHelper) {
    this.ticketChannelHelper = ticketChannelHelper;
  }

  @Async
  public CompletableFuture<MessageEmbed> buildTicketClosedEmbed(Ticket ticket) {
    try {
      final String ticketName = ticketChannelHelper.getTicketName(ticket).join();
      final Optional<User> author = ticket.getTicketAuthorNow();
      final Optional<User> closedBy = ticket.getClosedByNow();
      final String closeReason = ticket.getCloseReasonOrDefault();

      final StringBuilder descriptionBuilder = new StringBuilder("Ein Ticket wurde ");
      closedBy.ifPresent(user -> descriptionBuilder.append("von ").append(user.getAsMention()));
      descriptionBuilder.append("geschlossen.\n\n")
          .append("Grund: ")
          .append(closeReason);

      return CompletableFuture.completedFuture(new EmbedBuilder()
          .setTitle("Ticket \"%s\" geschlossen".formatted(ticketName))
          .setDescription(descriptionBuilder.toString())
          .setColor(EmbedColors.TICKET_CLOSED)
          .addField("Ticket-Id", toStringOrUnknown(ticket.getTicketId()), true)
          .addField("Ticket-Type", toStringOrUnknown(ticket.getTicketTypeString()), true)
          .addField("Ticket-Author", toStringOrUnknown(author.map(IMentionable::getAsMention)),
              true)
          .addField("Ticket-Eröffnungszeit", ticket.getOpenedAt().formatEuropeBerlin(), true)
          .addField("Ticket-Schließzeit", ticket.getClosedAt().formatEuropeBerlin(), true)
          .addField("Ticket-Dauer",
              ticket.getOpenedAt().formatEuropeBerlinDuration(ticket.getClosedAt()), true)
          .build());
    } catch (UnableToGetTicketNameException e) {
      return CompletableFuture.completedFuture(null);
    }
  }

  private String toStringOrUnknown(@Nullable Object object) {
    return object == null ? "Unbekannt" : object.toString();
  }

  private String toStringOrUnknown(@NotNull Optional<?> optional) {
    return optional.map(Object::toString).orElse("Unbekannt");
  }
}

File: src/main/kotlin/dev/slne/discord/message/EmbedManager.kt
package dev.slne.discord.message

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.messages.Embed
import dev.slne.discord.message.EmbedColors.TICKET_CLOSED
import dev.slne.discord.message.EmbedColors.TICKET_REOPENED
import dev.slne.discord.ticket.Ticket

object EmbedManager {

    fun buildTicketClosedUserPrivateMessageEmbed(
        ticket: Ticket,
        threadName: String? = ticket.thread?.name
    ) =
        Embed { // TODO: 14.10.2024 23:18 - use messages
            title = "Ticket \"${threadName ?: "Unbekannt"}\" geschlossen"

            description = buildString {
                append("Dein Ticket wurde geschlossen.\n\n")
                append("Grund: ${ticket.closeReasonOrDefault}\n\n")
                append("Weitere Informationen findest du im Ticket.\n${ticket.thread?.asMention}")
            }

            color = TICKET_CLOSED
        }

    suspend fun buildTicketClosedEmbed(ticket: Ticket, threadName: String? = ticket.thread?.name) =
        Embed { // TODO: 14.10.2024 23:18 - use messages
            title = "Ticket \"${threadName ?: "Unbekannt"}\" geschlossen"

            description = buildString {
                append("Ein Ticket wurde von ")
                ticket.closedBy?.await()?.let { append(it.asMention) }
                append(" geschlossen.\n\n")
                append("Grund: ${ticket.closeReasonOrDefault}")
            }

            color = TICKET_CLOSED

            field {
                name = "Ticket-Id"
                value = ticket.ticketId.toString()
            }

            field {
                name = "Ticket-Type"
                value = ticket.ticketType!!.displayName
            }

            field {
                name = "Ticket-Author"
                value = ticket.author?.await()?.asMention ?: "Unbekannt"
            }

            field {
                name = "Ticket-Eröffnungszeit"
                value = ticket.openedAt.format()
            }

            field {
                name = "Ticket-Schließzeit"
                value = ticket.closedAt.format()
            }

            field {
                name = "Ticket-Dauer"
                value = formatDuration(ticket.openedAt, ticket.closedAt)
            }
        }

    fun buildTicketReopenEmbed(ticket: Ticket) =
        Embed { // TODO: 14.10.2024 23:18 - use messages
            title = "Ticket \"${ticket.thread?.name ?: "Unbekannt"}\" wiedereröffnet"
            color = TICKET_REOPENED
        }
}
